catalog/dags/data_refresh/remap_table_constraints.py

```diff
@@ -0,0 +1,233 @@
+"""
+# Remap Table Constraints
+
+This file contains TaskGroups related to remapping the table constraints from the live media table onto the temp table to prepare it for promotion.
+
+"""
+
+import logging
+from typing import NamedTuple
+
+from airflow.decorators import task, task_group
+
+from common.sql import run_sql
+from data_refresh import queries
+
+
+logger = logging.getLogger(__name__)
+
+
+class ConstraintInfo(NamedTuple):
+    constraint_table: str
+    constraint_name: str
+    constraint_statement: str
+
+
+def fetch_all_tuples(cursor):
+    try:
+        rows = cursor.fetchall()
+        return [ConstraintInfo(row[0], row[1], row[2]) for row in rows]
+    except Exception as e:
+        raise ValueError("Unable to extract expected row data from cursor") from e
+
+
+def _is_foreign_key(_statement, table):
+    return f"REFERENCES {table}(" in _statement
+
+
+def _remap_constraint(
+    constraint_name: str,
+    constraint_table: str,
+    constraint_statement: str,
+    table_name: str,
+    temp_table_name: str,
+) -> list[str]:
+    """
+    Given information about an existing CONSTRAINT involving the original media table,
+    produce a list of `ALTER TABLE...` statements to remove the constraint from the
+    existing table and apply it to the temp table.
+
+    The constraint to remap may be a constraint directly on the live media table, in
+    which case it is dropped and then added directly to the temp table. Alternatively,
```

Comment:
Just to be clear, does that mean the live tables won't have those constraints for some time, even while they're still live? That's _probably_ fine, but if something fails part way through, could the constraints be missing on the live table the next time they're tried to be moved over?

Reply:
Correct. This _does_ mirror the current implementation, so it's not a regression and the (minor) improvement here is that we'd have immediate insight into where a failure happened/if those steps need to be manually undone.

I'll make an issue about this, because it's not simple to work around. We could do something similar to the way we remap the indices, where we give the new constraints a temporary name and then map the names back over in the promote step (if that's even necessary 🤔). But the `delete_orphans` step and remapping foreign key constraints to the temp table can't be easily postponed or undone.

I don't think we should be _too_ worried about this because it's the very last step before promotion, anyway -- but definitely good to call out, and something we'd _need_ to be aware of if for some reason we wanted to start the data refresh over from the beginning after a failure between this step and promotion.

Reply:
Thanks for the explanation, I think that sounds good! Not blocking in this case because it mirrors the previous implementation.